lib/services/local/generic/generic.go

```diff
@@ -64,9 +64,15 @@ type ServiceConfig[T Resource] struct {
 	// If set to 0, the default interval of 250ms will be used.
 	// WARNING: If set to a negative value, the RunWhileLocked function will retry immediately.
 	RunWhileLockedRetryInterval time.Duration
-	// KeyFunc optionally allows resource to have a custom key. If not provided the
-	// name of the resource will be used.
-	KeyFunc func(T) string
+	// NameKeyFunc optionally allows resources to have a custom key suffix, by
+	// transforming the name of the resource or the input given to methods that
+	// take a resource name (if ResourceKeyFunc is unset). If unset, the name is
+	// used without changes.
+	NameKeyFunc func(name string) string
+	// ResourceKeyFunc optionally allows resources to have a custom key suffix.
+	// If unset, the name of the resource is passed through NameKeyFunc if set,
+	// or used without changes otherwise.
+	ResourceKeyFunc func(T) string
```

Comment:
Do we really want to have two parameters that serve the same purpose of deducing the resource key?

What do you think about keeping only  the ResourceKeyFunc func(T) string function since it is the most generic approach and adding custom helper functions like:

```
func KeyNameWithSuffix[T types.ResourceMetadata](suffix string) func(T) string {
	return func(resource T) string {
		return resource.GetMetadata().GetName() + "/" + suffix
	}
}
```

```
func KeyCustomKeyName[T types.ResourceMetadata](fn func(string) string) func(T) string {
	return func(resource T) string {
		return fn(resource.GetMetadata().GetName())
	}
}
```

```
func StaticResourceKey[T types.ResourceMetadata](s string) func(T) string {
	return func(resource T) string {
		return s
	}
}
```

Reply:
I considered doing the opposite and only keeping the function that transforms the resource name. The problem that this PR is trying to solve is that if you only have a way to redefine the key used with a resource you can't do anything if all you have is a name, which happens with Get or Delete.

You can sort of get around it by calling the methods of the generic service with a pre-transformed name, but then the errors generated by the generic service (like "not found") will have the transformed name in it so you also have to rewrite those.

Reply:
I explained myself poorly - I considered only leaving `NameKeyFunc` and using it with the return value of `resource.GetName()` for every function that has a resource, but I didn't want to get rid of the current level of flexibility (which is not used much, but still).

Reply:
Actually, the only user of `ResourceKeyFunc` is three singleton resources that are currently using a const name for Get and Delete anyway, and I think that for `ResourceKeyFunc` to actually be flexible enough for other purposes it would have to return a `backend.Key` or a `[]string` or something, so I'm just going to get rid of that and only leave `NameKeyFunc` in.